### PR TITLE
Resolved Issue #1170

### PR DESCRIPTION
**Pull request by shivsahni**
The content for device binding using encryption methodology is revised to prevent any ambiguities  that could lead to insecure sensitive data storage.

This PR covers issue #1170

### CODE REVIEW

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 1]
```diff
@@ -1241,7 +1241,19 @@ The goal of device binding is to impede an attacker who tries to both copy an ap
...
-- Obfuscating the data stored on the device by using device identifiers as keys for encryption methods. This can help with binding to a device when the app does a lot of offline work or when access to APIs depends on access-tokens stored by the application.
+
+- Encrypting the data stored in the device with the key material which is strongly bound to the device can help in the implementation of the device binding. The Android Keystore can be leveraged for the same to generate and manage the key material which is strongly coupled with the device. This would prevent the malicious actor to perform operations from the device to which the data is copied since the key material with which the data is encrypted would not be present to decrypt the encrypted data. This can be implemented using the following way:
+
+  - Generating the key pair in the Android keystore using *KeyPairGeneratorSpec* API.
+  - Generating a secret key for AES-GCM using a Secure Pseudo Random Number Generator(SPRNG) such as *java.security.SecureRandom* API
```
Comment (commjoen) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: The *java.security.SecureRandom* is not a SPRNG, it is a wrapper using secure random sources, rather say "through APIs such as the *java.security.SecureRandom* API"

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 2]
```diff
@@ -1241,7 +1241,19 @@ The goal of device binding is to impede an attacker who tries to both copy an ap
...
+
+- Encrypting the data stored in the device with the key material which is strongly bound to the device can help in the implementation of the device binding. The Android Keystore can be leveraged for the same to generate and manage the key material which is strongly coupled with the device. This would prevent the malicious actor to perform operations from the device to which the data is copied since the key material with which the data is encrypted would not be present to decrypt the encrypted data. This can be implemented using the following way:
+
+  - Generating the key pair in the Android keystore using *KeyPairGeneratorSpec* API.
+  - Generating a secret key for AES-GCM using a Secure Pseudo Random Number Generator(SPRNG) such as *java.security.SecureRandom* API
+  - Encrypt the authentication data and other sensitive data stored by the application using secret key through AES-GCM cipher
+  - Encrypt the secret key using public key stored in Android keystore and store the same in the private storage of the application
```
Comment (commjoen) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: What do you mean with the same here :) ?

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 3]
```diff
@@ -1241,7 +1241,19 @@ The goal of device binding is to impede an attacker who tries to both copy an ap
...
+  - Whenever authentication data or other sensitive data is required, decrypt the secret key using private key stored in Android keystore and then use the same to decrypt the ciphertext
+
+  Note: For API level 23 and above, the *KeyGenParameterSpec* API can be leveraged directly to generate and manage secret key through Android keystore.
```
Comment (TheDauntless) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: > ... to generate and manage secret **key** through Android keystore.

key -> keys

**Comment by commjoen** [comment]
Can you please, before you fix the feedback, merge master in?

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 4]
```diff
@@ -1241,7 +1241,19 @@ The goal of device binding is to impede an attacker who tries to both copy an ap
...
+
+  For the detailed implementation guidelines, [Securely Storing Secrets in an Android Application](https://medium.com/@ericfu/securely-storing-secrets-in-an-android-application-501f030ae5a3 "Securely Storing Secrets in an Android Application") blog can be referred.
```
[the remark this answers is not in the document]
Comment (shivsahni) [reply Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Removed

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 5]
```diff
@@ -1486,6 +1498,10 @@ For a more detailed assessment, you need a detailed understanding of the relevan
 - SafetyNet Verification Samples - https://github.com/googlesamples/android-play-safetynet/
 - SafetyNet Attestation API - Quota Request - https://support.google.com/googleplay/android-developer/contact/safetynetqr
 
+#### Secure Storage and Device Binding
+
+- Securely Storing Secrets in an Android Application - https://medium.com/@ericfu/securely-storing-secrets-in-an-android-application-501f030ae5a3
```
[the remark this answers is not in the document]
Comment (shivsahni) [reply Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Removed.

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md (hunk 2 above)
Comment (shivsahni) [reply Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Yes, made that more descriptive now.

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 6]
```diff
@@ -1237,7 +1237,17 @@ The goal of device binding is to impede an attacker who tries to both copy an ap
 Before we describe the usable identifiers, let's quickly discuss how they can be used for binding. There are three methods that allow device binding:
```
Comment (TheDauntless) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: We now have more than three methods.

Comment (shivsahni) [reply Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: Hi @TheDauntless,

I believe we still have three methods listed in the context. I have just elaborated and removed the scope of any ambiguities on one point that was already mentioned.

Comment (TheDauntless) [reply Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md]: My bad, didn't see the indent of the 2nd list :).

**Comment by shivsahni** [comment]
@TheDauntless, is there anything pending. I think we are good to go and this can be merged.

**Pull request merged**